### PR TITLE
Ported OSAL from most recent cFS version

### PR DESCRIPTION
**Pull request by PaulaBosca**
Changes include incorporating Uzair's porting along with some of my own (PaulaBosca)

**Describe the contribution**
Ported all functions that required it (header files were not touched, alongside some files which did not have any implemented functions. 
-Werror in the build file was commented out as it treated warning as errors, and all warning were either unused function or uninitialized variables (but still declared)). 

**Testing performed**
In main cFS directory, used 'make', with -Werror turned off. 

**System(s) tested on**
Ubuntu:Bionic (18.04)

**Code contributions**
Uzair Abbasi's porting work was the majority of the edits.

### CODE REVIEW

**Comment by PaulaBosca** [comment]
For those reviewing, a couple questions/concerns/things to please check out: 

In the file os-impl-tasks.c 
- Function OS_TaskGetId_Impl on line 549 was not edited as I wasn't sure how 
- Please double-check function OS_TaskIdMatchSystemData_Impl on line 500

In file os-impl-common.c 
- Function OS_IdleLoop_Impl on line 75 has a void return, I wasn't sure if the function contents should be removed
- Function OS_ApplicationShutdown_Impl on line 97 also has a void return, same as above
- Function OS_Posix_CompAbsDelayTime on line 117 also has a void return and this time i did just delete everything. Please let me know if this is incorrect

In file os-impl-timebase.c 
- Function OS_UsecToTimespec on line 90 has a void return and none of the contents seemed OS-specific (although I think some may be part of OS-specific structures, unsure). I left this one alone. Please double check. 
- Function OS_TimeBaseLock_Impl on line 113 and OS_TimeBaseUnlock_Impl on line 126 are return void with no contents

File: osal/src/os/posix/inc/os-posix.h [hunk 1]
```diff
@@ -54,7 +54,7 @@
 /*
  * Use the global definitions from the shared layer
  */
-#include <os-shared-globaldefs.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-posix.h:57]: Is this change necessary?

**Comment by JohnBarberiz** [comment]
Probably not. There is a bit left to do on this. Alot of the linux includes
were left in as well. I'll go through it tomorrow and see if I can fix it
up.

File: osal/src/bsp/generic-linux/src/bsp_console.c [hunk 2]
```diff
@@ -30,39 +30,8 @@
 #include <sys/types.h>
 #include <sys/wait.h>
```
Comment (stephen99scott) [inline osal/src/bsp/generic-linux/src/bsp_console.c:31]: Posix dependencies should be removed

File: osal/src/os/posix/inc/os-freertos.h [hunk 3]
```diff
@@ -40,21 +40,16 @@
 #include <stdlib.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-freertos.h:40]: This was removed from `bsp_start.c`, should it be removed here, or added back there?

File: osal/src/os/posix/inc/os-freertos.h [hunk 4]
```diff
@@ -40,21 +40,16 @@
...
 #include <string.h>
 #include <unistd.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-freertos.h:42]: Posix dependency should be removed

File: osal/src/os/posix/inc/os-freertos.h [hunk 5]
```diff
@@ -40,21 +40,16 @@
...
 #include <signal.h>
 #include <time.h>
 #include <errno.h>
-#include <pthread.h>
-#include <mqueue.h>
-#include <fcntl.h>
-#include <semaphore.h>
 #include <sys/types.h>
-#include <sys/signal.h>
+#include <signal.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-freertos.h:47]: This is already included on line 43

File: osal/src/os/posix/inc/os-freertos.h [hunk 6]
```diff
@@ -40,21 +40,16 @@
...
 
 /*
  * Use the global definitions from the shared layer
  */
-#include <os-shared-globaldefs.h>
+#include "os-shared-globaldefs.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-freertos.h:52]: I don't think this should be altered

File: osal/src/os/posix/inc/os-freertos.h [hunk 7]
```diff
@@ -40,21 +40,16 @@
...
 #include <unistd.h>
-#include <limits.h>
 #include <signal.h>
 #include <time.h>
 #include <errno.h>
-#include <pthread.h>
-#include <mqueue.h>
-#include <fcntl.h>
-#include <semaphore.h>
 #include <sys/types.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-freertos.h:46]: Posix dependency should be removed

File: osal/src/os/posix/inc/os-freertos.h [hunk 8]
```diff
@@ -40,21 +40,16 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
-#include <limits.h>
 #include <signal.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-freertos.h:43]: Are these headers implemented in freeRTOS?

File: osal/src/os/posix/inc/os-impl-io.h [hunk 9]
```diff
@@ -20,33 +20,29 @@
...
  *
  */
 
 #ifndef INCLUDE_OS_IMPL_IO_H_
 #define INCLUDE_OS_IMPL_IO_H_
 
 #include <osconfig.h>
-#include <common_types.h>
+#include <stdbool.h>
 #include <unistd.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-impl-io.h:32]: Posix dependency should be removed

File: osal/src/os/posix/inc/os-impl-io.h [hunk 10]
```diff
@@ -20,33 +20,29 @@
...
+ * \ingroup  FreeRTOS
  *
  */
 
 #ifndef INCLUDE_OS_IMPL_IO_H_
 #define INCLUDE_OS_IMPL_IO_H_
 
 #include <osconfig.h>
-#include <common_types.h>
+#include <stdbool.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-impl-io.h:31]: Why was this change made?

File: osal/src/os/posix/inc/os-impl-sockets.h [hunk 11]
```diff
@@ -31,12 +30,9 @@
 #include "os-impl-io.h"
 
 #include <fcntl.h>
-#include <netinet/in.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/inc/os-impl-sockets.h:34]: Posix dependencies should be removed

File: osal/src/os/posix/src/os-impl-common.c [hunk 12]
```diff
@@ -20,18 +20,14 @@
...
+ * \ingroup  FreeRTOS
  *
  */
 
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-#include "os-posix.h"
-#include "bsp-impl.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-common.c:33]: Is it necessary to remove this include?

File: osal/src/os/posix/src/os-impl-dirs.c [hunk 13]
```diff
@@ -32,11 +31,10 @@
 #include <fcntl.h>
 #include <dirent.h>
 #include <sys/stat.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-dirs.c:33]: Posix dependencies should be removed

File: osal/src/os/posix/src/os-impl-dirs.c [hunk 14]
```diff
@@ -32,11 +31,10 @@
...
-#include <sys/types.h>
-#include <sys/wait.h>
 
-#include "os-posix.h"
 #include "os-impl-dirs.h"
+
+#include "os-freertos.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-dirs.c:37]: There were other locations where `os-posix.h` was removed but `os-freertos.h` was not added in those instances, should we add the freeRTOS include to those locations as well?

File: osal/src/os/posix/src/os-impl-errors.c [hunk 15]
```diff
@@ -20,17 +20,14 @@
...
- * \author   joseph.p.hickey@nasa.gov
+ * \ingroup  FreeRTOS
  *
  */
 
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-#include "os-posix.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-errors.c:32]: e.g. should we add `os-freertos.h` here?

File: osal/src/os/posix/src/os-impl-filesys.c [hunk 16]
```diff
@@ -20,30 +20,19 @@
...
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-
-#include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
-#include <sys/types.h>
 #include <fcntl.h>
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-filesys.c:31]: Posix dependency should be removed

File: osal/src/os/posix/src/os-impl-fpu.c [hunk 17]
```diff
@@ -20,17 +20,14 @@
...
+ * \ingroup  FreeRTOS
  *
  */
 
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-#include "os-posix.h"
-#include "bsp-impl.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-fpu.c:33]: Is it necessary to remove this include?

File: osal/src/os/posix/src/os-impl-heap.c [hunk 18]
```diff
@@ -20,16 +20,14 @@
...
- * \author   joseph.p.hickey@nasa.gov
+ * \ingroup  FreeRTOS
  *
  */
 
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-#include "os-posix.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-heap.c:32]: Should we include `os-freertos.h`?

File: osal/src/os/posix/src/os-impl-idmap.c [hunk 19]
```diff
@@ -20,61 +20,17 @@
...
+ * \ingroup  FreeRTOS
  *
  */
 
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-#include "os-posix.h"
-#include "bsp-impl.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-idmap.c:33]: Should this be removed?

File: osal/src/os/posix/src/os-impl-no-module.c [hunk 20]
```diff
@@ -20,20 +20,15 @@
...
 
 /****************************************************************************************
                                     INCLUDE FILES
  ***************************************************************************************/
 
-#include <stdlib.h>
-#include <string.h>
-
 #include <osapi.h>
-#include "os-shared-module.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-no-module.c:36]: Should this be removed?

File: osal/src/os/posix/src/os-impl-shell.c [hunk 21]
```diff
@@ -20,25 +20,17 @@
...
-#include <fcntl.h>
-#include <sys/stat.h>
-#include <sys/types.h>
-#include <sys/wait.h>
-
-#include "os-posix.h"
 #include "os-impl-io.h"
 
 #include "os-shared-shell.h"
-#include "os-shared-idmap.h"
```
Comment (stephen99scott) [inline osal/src/os/posix/src/os-impl-shell.c:41]: Is this necessary to remove?

File: osal/src/bsp/generic-linux/src/bsp_console.c (hunk 2 above)
Comment (JohnBarberiz) [reply osal/src/bsp/generic-linux/src/bsp_console.c:31]: I don't mind removing these, I left them in because they were left in for the rtems version as well for some reason which is what I used to compare.
Screen Shot 2020-08-24 at 6 22 47 PM

Comment (JohnBarberiz) [reply osal/src/bsp/generic-linux/src/bsp_console.c:31]: Thinking it over I think its best its removed. Its possible when rtems is built that they built it on a linux host which is how those are defined. But for rtems they are removed.

File: osal/src/os/posix/inc/os-freertos.h (hunk 3 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/inc/os-freertos.h:40]: I think either is fine. My general reasoning was I diffed the posix directory with either vxworks/rtems (forget which) and if it still had the include I left it there. I figured at the team it was best to leave so people knew what libraries were available. But we can just strip out everything that still lets it compile.

File: osal/src/os/posix/inc/os-freertos.h (hunk 4 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/inc/os-freertos.h:42]: These are also in the rtems/vxworks builds but again no harm in removing them. But we are likely to use them in that file since every version of it has these files.

File: osal/src/os/posix/inc/os-freertos.h (hunk 5 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/inc/os-freertos.h:47]: that's a goof.

File: osal/src/os/posix/inc/os-freertos.h (hunk 6 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/inc/os-freertos.h:52]: Yeah I think its fine leaving it how it was. generally the quotation marks implies the header can be found within the project, which is what is happening in this case. but yeah can revert this

File: osal/src/os/posix/src/os-impl-fpu.c (hunk 17 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/src/os-impl-fpu.c:33]: the posix which was renamed should stay. The bsp-impl should be removed, rtems version at least does not include it.

File: osal/src/os/posix/src/os-impl-heap.c (hunk 18 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/src/os-impl-heap.c:32]: yeah probably should be left in.

File: osal/src/os/posix/src/os-impl-idmap.c (hunk 19 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/src/os-impl-idmap.c:33]: posix should be left in, bsp-impl removed as per previous comment.

File: osal/src/os/posix/src/os-impl-no-module.c (hunk 20 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/src/os-impl-no-module.c:36]: yeah, again rtems doesnt have it, but they don't implement anything in this function either

File: osal/src/os/posix/src/os-impl-shell.c (hunk 21 above)
Comment (JohnBarberiz) [reply osal/src/os/posix/src/os-impl-shell.c:41]: same as previous comment

File: osal/src/os/posix/inc/os-freertos.h (hunk 6 above)
Comment (PaulaBosca) [reply osal/src/os/posix/inc/os-freertos.h:42]: yea that was my bad, I was playing around with header files and forgot to change it back.

**Review by stephen99scott — APPROVED** [review]
* I think we should create a new directory for these files, but I am approving so we can move forward with porting

**Comment by stephen99scott** [comment]
Creating a pull request with the stubbed code in a new directory
